server/src/main/java/keywhiz/service/providers/ClientAuthFactory.java

```diff
@@ -34,58 +44,253 @@
     return Optional.ofNullable(request.getSecurityContext().getUserPrincipal());
   }
 
-  static Optional<String> getClientName(Optional<Principal> principal) {
-    if (!principal.isPresent()) {
-      return Optional.empty();
-    }
-
-    X500Name name = new X500Name(principal.get().getName());
+  static Optional<String> getClientName(Principal principal) {
+    X500Name name = new X500Name(principal.getName());
     RDN[] rdns = name.getRDNs(BCStyle.CN);
     if (rdns.length == 0) {
-      logger.warn("Certificate does not contain CN=xxx,...: {}", principal.get().getName());
+      logger.warn("Certificate does not contain CN=xxx,...: {}", principal.getName());
       return Optional.empty();
     }
     return Optional.of(IETFUtils.valueToString(rdns[0].getFirst().getValue()));
   }
 
+  private Optional<XfccSourceConfig> getXfccConfigForPort(int port) {
+    List<XfccSourceConfig> matchingConfigs = clientAuthConfig.xfccConfigs()
+        .stream()
+        .filter(xfccSourceConfig -> xfccSourceConfig.port().equals(port))
+        .collect(Collectors.toUnmodifiableList());
+
+    if (matchingConfigs.size() > 1) {
+      throw new NotAuthorizedException(format(
+          "Invalid 'xfcc' configuration for port %d; at most one configuration must be present per port",
+          port));
+    } else {
+      return matchingConfigs.stream().findFirst();
+    }
+  }
+
+  private Client authorizeClientFromXfccHeader(XfccSourceConfig xfccConfig,
+      List<String> xfccHeaderValues, Optional<String> requestName,
+      Optional<String> requestSpiffeId) {
+    // Do not allow the XFCC header to be set by all incoming traffic. This throws a
+    // NotAuthorizedException when the traffic is not coming from a source allowed to set the
+    // header.
+    validateXfccHeaderAllowed(xfccConfig, requestName, requestSpiffeId);
+
+    // Extract client information from the XFCC header
+    Optional<X509Certificate> clientCert =
+        getClientCertFromXfccHeaderEnvoyFormatted(xfccHeaderValues);
+    if (clientCert.isEmpty()) {
+      throw new NotAuthorizedException(
+          format("unable to parse client certificate from %s header", XFCC_HEADER_NAME));
+    }
+
+    Principal clientPrincipal = clientCert.get().getSubjectX500Principal();
+    Optional<String> clientName = getClientName(clientPrincipal);
+    Optional<String> clientSpiffeId = Optional.empty();
+
+    return authorizeClient(clientPrincipal, clientName, clientSpiffeId);
+  }
+
+  private void validateXfccHeaderAllowed(XfccSourceConfig xfccConfig, Optional<String> requestName,
+      Optional<String> requestSpiffeId) {
+    if (clientAuthConfig == null || clientAuthConfig.xfccConfigs() == null) {
+      throw new NotAuthorizedException(
+          format(
+              "not configured to handle requests with %s header set; set 'xfcc' field in config",
+              XFCC_HEADER_NAME));
+    }
+
+    if (requestName.isEmpty() && requestSpiffeId.isEmpty()) {
```

Comment:
isEmpty() -- if Keywhiz could parse neither a CN nor a SPIFFE ID from the entity making this request, which set the XFCC header, then it can't check that entity against the allowlist for setting the header and it fails closed. 

If either the CN or the SPIFFE ID were set, then Keywhiz checks that identity against the corresponding allowlist to make sure the entity's allowed to set the XFCC header.